common/lib/xmodule/xmodule/modulestore/split_mongo/split.py

```diff
@@ -2045,6 +2044,157 @@ def copy(self, user_id, source_course, destination_course, subtree_list=None, bl
             self.update_structure(destination_course, destination_structure)
             self._update_head(destination_course, index_entry, destination_course.branch, destination_structure['_id'])
 
+    @contract(source_keys="list(BlockUsageLocator)", dest_usage=BlockUsageLocator)
+    def inherit_copy(self, source_keys, dest_usage, user_id, copy_children=True):
+        """
+        Flexible mechanism for inheriting content from an external course/library/etc.
+
+        Will copy all of the XBlocks whose keys are passed as `source_course` so that they become
+        children of the XBlock whose key is `dest_usage`. Any previously existing children of
+        `dest_usage` that haven't been replaced/updated by this inherit_copy operation will be
+        deleted.
+
+        Unlike `copy()`, this does not care whether the resulting blocks are positioned similarly
+        in their new course/library. However, the resulting blocks will be in the same relative
+        order as `source_keys`.
+
+        If any of the blocks specified already exist as children of the destination block, they
+        will be updated rather than duplicated or replaced. If they have Scope.settings field values
+        overriding inherited default values, those overrides will be preserved.
+
+        IMPORTANT: This method does not preserve block_id - in other words, every block that is
+        copied will be assigned a new block_id. This is because we assume that the same source block
+        may be copied into one course in multiple places. However, it *is* guaranteed that every
+        time this method is called for the same source block and dest_usage, the same resulting
+        block id will be generated.
+
+        :param source_keys: a list of BlockUsageLocators. Order is preserved.
+
+        :param dest_usage: The BlockUsageLocator that will become the parent of an inherited copy
+        of all the xblocks passed in `source_keys`.
+
+        :param user_id: The user who will get credit for making this change.
+
+        :param copy_children: If true, all descendants of each XBlock will be copied recursively.
+        """
+        if copy_children:
+            # Preload the block structures for all source courses/libraries/etc.
+            # so that we can access descendant information quickly
+            source_structures = {}
+            for key in source_keys:
+                course = key.course_key.for_version(None)
+                if course.branch is None:
+                    raise ItemNotFoundError("branch is required for all source keys when using inherit_copy")
+                if course not in source_structures:
+                    with self.bulk_operations(course):
```

Comment:
I just tried moving this up like you suggested but it doesn't seem to work without a valid course key. Also, 99% of the time, all of the source_keys will be coming from just one library/course, so it's only going to be calling `_lookup_course` once anyways. Thus I'm going to leave it for now - let me know if you want me to revisit.